File: apps/backend/payments/money.py
"""Money & Currency — the server mirror of the Flutter client's value objects
(`apps/mobile/lib/features/wallet/domain/money.dart` + `currency.dart`).

Value is stored as an integer count of **minor units** in a single currency.
Arithmetic is exact; `float`/`Decimal` are never used for value so balances,
fees and ledger postings can never drift. Keeping this identical to the client
is what lets the two stay aligned by construction.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from functools import total_ordering


class Currency(Enum):
    # code, symbol, name, minor_unit_digits, display_digits, is_crypto
    TZS = ("TZS", "TSh", "Tanzanian Shilling", 2, 0, False)
    USD = ("USD", "$", "US Dollar", 2, 2, False)
    EUR = ("EUR", "€", "Euro", 2, 2, False)
    KES = ("KES", "KSh", "Kenyan Shilling", 2, 0, False)
    UGX = ("UGX", "USh", "Ugandan Shilling", 0, 0, False)
    RWF = ("RWF", "FRw", "Rwandan Franc", 0, 0, False)
    BIF = ("BIF", "FBu", "Burundian Franc", 0, 0, False)
    ZMW = ("ZMW", "ZK", "Zambian Kwacha", 2, 2, False)
    MWK = ("MWK", "MK", "Malawian Kwacha", 2, 2, False)
    CDF = ("CDF", "FC", "Congolese Franc", 2, 2, False)
    BTC = ("BTC", "₿", "Bitcoin", 8, 6, True)

    def __init__(self, code, symbol, name, minor_unit_digits, display_digits, is_crypto):
        self.code = code
        self.symbol = symbol
        self.currency_name = name
        self.minor_unit_digits = minor_unit_digits
        self.display_digits = display_digits
        self.is_crypto = is_crypto

    @property
    def scale(self) -> int:
        return 10 ** self.minor_unit_digits

    @classmethod
    def from_code(cls, code: str) -> "Currency":
        for c in cls:
            if c.code == code.upper():
                return c
        raise ValueError(f"Unsupported currency: {code}")


@total_ordering
@dataclass(frozen=True)
class Money:
    """An immutable monetary amount in integer minor units."""

    minor_units: int
    currency: Currency

    @classmethod
    def major(cls, major: int, currency: Currency) -> "Money":
        return cls(int(major) * currency.scale, currency)

    @classmethod
    def from_decimal(cls, amount: float, currency: Currency) -> "Money":
        # Round half-up to the currency's minor-unit precision.
        return cls(int(round(amount * currency.scale)), currency)

    @classmethod
    def zero(cls, currency: Currency) -> "Money":
        return cls(0, currency)

    # --- properties ---
    @property
    def is_zero(self) -> bool:
        return self.minor_units == 0

    @property
    def is_positive(self) -> bool:
        return self.minor_units > 0

    @property
    def is_negative(self) -> bool:
        return self.minor_units < 0

    @property
    def as_decimal(self) -> float:
        return self.minor_units / self.currency.scale

    @property
    def abs(self) -> "Money":
        return Money(abs(self.minor_units), self.currency)

    # --- arithmetic (same-currency only) ---
    def _assert_same(self, other: "Money") -> None:
        if self.currency != other.currency:
            raise ValueError(
                f"Currency mismatch: {self.currency.code} vs {other.currency.code}. "
                "Convert via the currency engine before combining."
            )

    def __add__(self, other: "Money") -> "Money":
        self._assert_same(other)
        return Money(self.minor_units + other.minor_units, self.currency)

    def __sub__(self, other: "Money") -> "Money":
        self._assert_same(other)
        return Money(self.minor_units - other.minor_units, self.currency)

    def __neg__(self) -> "Money":
        return Money(-self.minor_units, self.currency)

    def __lt__(self, other: "Money") -> bool:
        self._assert_same(other)
        return self.minor_units < other.minor_units

    # __eq__ + frozen dataclass gives value equality; total_ordering fills the rest.

    def format(self, with_symbol: bool = True, with_sign: bool = False) -> str:
        negative = self.minor_units < 0
        scaled = abs(self.minor_units)
        major = scaled // self.currency.scale
        out = f"{major:,}"
        if self.currency.display_digits > 0:
            padded = str(scaled % self.currency.scale).rjust(self.currency.minor_unit_digits, "0")
            out += "." + padded[: self.currency.display_digits].ljust(self.currency.display_digits, "0")
        sign = "-" if negative else ("+" if with_sign and self.minor_units > 0 else "")
        symbol = f"{self.currency.symbol} " if with_symbol else ""
        return f"{sign}{symbol}{out}"

    def __str__(self) -> str:
        return f"{self.currency.code} {self.format(with_symbol=False)}"

File: apps/mobile/lib/features/wallet/domain/money.dart
import 'currency.dart';

/// An immutable monetary amount stored as **integer minor units** in a single
/// [Currency]. Money math is exact — we never use `double` for value, which
/// eliminates rounding drift in balances, fees and ledger postings.
///
/// Example: `Money.major(2847500, Currency.tzs)` is TSh 2,847,500.00 stored as
/// 284,750,000 minor units.
class Money implements Comparable<Money> {
  const Money(this.minorUnits, this.currency);

  /// The signed amount in minor units (e.g. cents).
  final int minorUnits;
  final Currency currency;

  /// Build from a whole major-unit amount (e.g. shillings).
  factory Money.major(int major, Currency currency) =>
      Money(major * currency.scale, currency);

  /// Build from a decimal major amount (e.g. 100.20 USD). Rounds half-up to the
  /// currency's minor-unit precision.
  factory Money.fromDecimal(num amount, Currency currency) {
    final scaled = amount * currency.scale;
    return Money(scaled.round(), currency);
  }

  static Money zero(Currency currency) => Money(0, currency);

  bool get isZero => minorUnits == 0;
  bool get isNegative => minorUnits < 0;
  bool get isPositive => minorUnits > 0;

  /// Value expressed in major units as a double — **for display/estimates only**,
  /// never for further money math.
  double get asDecimal => minorUnits / currency.scale;

  Money operator +(Money other) {
    _assertSameCurrency(other);
    return Money(minorUnits + other.minorUnits, currency);
  }

  Money operator -(Money other) {
    _assertSameCurrency(other);
    return Money(minorUnits - other.minorUnits, currency);
  }

  Money operator -() => Money(-minorUnits, currency);

  Money get abs => Money(minorUnits.abs(), currency);

  bool operator >(Money other) {
    _assertSameCurrency(other);
    return minorUnits > other.minorUnits;
  }

  bool operator <(Money other) {
    _assertSameCurrency(other);
    return minorUnits < other.minorUnits;
  }

  bool operator >=(Money other) => this > other || this == other;
  bool operator <=(Money other) => this < other || this == other;

  void _assertSameCurrency(Money other) {
    if (other.currency != currency) {
      throw StateError(
        'Currency mismatch: ${currency.code} vs ${other.currency.code}. '
        'Convert via the currency engine before combining.',
      );
    }
  }

  /// Formats with grouping separators and the currency symbol, e.g.
  /// `TSh 2,847,500` or `\$1,140.00`.
  String format({bool withSymbol = true, bool withSign = false}) {
    final negative = minorUnits < 0;
    final scaled = minorUnits.abs();
    final major = scaled ~/ currency.scale;
    final fraction = scaled % currency.scale;

    final buffer = StringBuffer(_group(major));
    if (currency.displayDigits > 0) {
      // Right-most `displayDigits` of the zero-padded minor part.
      final padded = fraction.toString().padLeft(currency.minorUnitDigits, '0');
      final shown = currency.displayDigits <= padded.length
          ? padded.substring(0, currency.displayDigits)
          : padded.padRight(currency.displayDigits, '0');
      buffer.write('.$shown');
    }

    final sign = negative ? '-' : (withSign && minorUnits > 0 ? '+' : '');
    final symbol = withSymbol ? '${currency.symbol} ' : '';
    return '$sign$symbol${buffer.toString()}';
  }

  static String _group(int value) {
    final s = value.toString();
    final b = StringBuffer();
    for (var i = 0; i < s.length; i++) {
      if (i > 0 && (s.length - i) % 3 == 0) b.write(',');
      b.write(s[i]);
    }
    return b.toString();
  }

  @override
  int compareTo(Money other) {
    _assertSameCurrency(other);
    return minorUnits.compareTo(other.minorUnits);
  }

  @override
  bool operator ==(Object other) =>
      other is Money &&
      other.minorUnits == minorUnits &&
      other.currency == currency;

  @override
  int get hashCode => Object.hash(minorUnits, currency);

  @override
  String toString() => '${currency.code} ${format(withSymbol: false)}';
}
